msg_encoding_test.go

```diff
@@ -0,0 +1,161 @@
+package f3
+
+import (
+	"math/rand"
+	"testing"
+
+	"github.com/filecoin-project/go-bitfield"
+	"github.com/filecoin-project/go-f3/gpbft"
+	"github.com/ipfs/go-cid"
+	"github.com/multiformats/go-multihash"
+	"github.com/stretchr/testify/require"
+)
+
+const seed = 1413
+
+func BenchmarkCborEncoding(b *testing.B) {
+	rng := rand.New(rand.NewSource(seed))
+	encoder := &cborGMessageEncoding{}
+	msg := generateRandomGMessage(b, rng)
+
+	b.ResetTimer()
+	b.ReportAllocs()
+	b.RunParallel(func(pb *testing.PB) {
+		for pb.Next() {
+			if _, err := encoder.Encode(msg); err != nil {
+				require.NoError(b, err)
+			}
+		}
+	})
+}
+
+func BenchmarkCborDecoding(b *testing.B) {
+	rng := rand.New(rand.NewSource(seed))
+	encoder := &cborGMessageEncoding{}
+	msg := generateRandomGMessage(b, rng)
+	data, err := encoder.Encode(msg)
+	require.NoError(b, err)
+
+	b.ResetTimer()
+	b.ReportAllocs()
+	b.RunParallel(func(pb *testing.PB) {
+		for pb.Next() {
+			if got, err := encoder.Decode(data); err != nil {
+				require.NoError(b, err)
+				require.Equal(b, msg, got)
+			}
+		}
+	})
+}
+
+func BenchmarkZstdEncoding(b *testing.B) {
+	rng := rand.New(rand.NewSource(seed))
+	encoder, err := newZstdGMessageEncoding()
+	require.NoError(b, err)
+	msg := generateRandomGMessage(b, rng)
+
+	b.ResetTimer()
+	b.ReportAllocs()
+	b.RunParallel(func(pb *testing.PB) {
+		for pb.Next() {
+			if _, err := encoder.Encode(msg); err != nil {
+				require.NoError(b, err)
+			}
+		}
+	})
+}
+
+func BenchmarkZstdDecoding(b *testing.B) {
+	rng := rand.New(rand.NewSource(seed))
+	encoder, err := newZstdGMessageEncoding()
+	require.NoError(b, err)
+	msg := generateRandomGMessage(b, rng)
+	data, err := encoder.Encode(msg)
+	require.NoError(b, err)
+
+	b.ResetTimer()
+	b.ReportAllocs()
+	b.RunParallel(func(pb *testing.PB) {
+		for pb.Next() {
+			if got, err := encoder.Decode(data); err != nil {
+				require.NoError(b, err)
+				require.Equal(b, msg, got)
+			}
+		}
+	})
+}
+
+func generateRandomGMessage(b *testing.B, rng *rand.Rand) *gpbft.GMessage {
+	return &gpbft.GMessage{
+		Sender:        gpbft.ActorID(rng.Uint64()),
+		Vote:          generateRandomPayload(b, rng),
+		Signature:     generateRandomBytes(b, rng, 96),
+		Ticket:        generateRandomBytes(b, rng, rng.Intn(96)+16),
```

Comment:
Ticket size is constant AFAIK

Reply:
Changed to fixed size bimodal presence at random (ignoring message validity for now since it doesn't matter for benchmarking)